packages/material-ui-styles/src/withStyles/withStyles.js

```diff
@@ -67,6 +68,25 @@ const withStyles = (stylesOrCreator, options = {}) => (Component) => {
       if (withTheme && !more.theme) {
         more.theme = theme;
       }
+
+      if (theme && theme.variants && theme.variants[name]) {
```

Comment:
@oliviertassinari here is the logic for the dynamic classes.

Reply:
Considering this failing test case:

```jsx
    it.only('should map the variant classkey to the component', function test() {
      if (/jsdom/.test(window.navigator.userAgent)) {
        // see https://github.com/jsdom/jsdom/issues/2953
        this.skip();
      }

      const theme = createMuiTheme({
        variants: {
          MuiButton: [
            {
              props: { variant: 'test', color: 'primary', size: 'large' },
              styles: { backgroundColor: 'rgb(255, 0, 0)' },
            },
          ],
        },
      });

      render(<WrappedComponent theme={theme} variant="test" size="large" />);

      const style = window.getComputedStyle(screen.getByTestId('component'));
      expect(style.getPropertyValue('background-color')).to.equal('rgb(255, 0, 0)');
    });
```

Should we move the logic outside of withStyles, after the resolution of the default props?

Reply:
Let me test this out a bit... The defaults for this are set inside the render method of the component, so we can potentially move it there, but that would mean that it needs to be added as in each component separately (which won't be a big deal as it can be implemented as a hook). Let me test this tomorrow and ensure that it works

Reply:
Related question, why are we using sometimes the static `defailtProps`, and other times defaults in the render method (the second should be better as `defaultProps` will be removed`), but it is confusing - I thought initially that it is default to spread the defaultProps there, as it was implemented like that previosly, but that is obviously not enough.